README.md

```diff
@@ -1,30 +1,205 @@
+    country: null
+  },
+}
+```
+
+If we wanted to update the `addressInfo.city` field, you might think we can update it like this:
+
+```js
+this.setState({
+  address: {
+    city: 'New York City',
+  },
+});
+```
+
+**However**, this would result in the following state shape:
+
+```
+{
+  theme: 'blue',
+  addressInfo: {
+    city: 'New York City',
+  },
+}
+```
 
+See what happened there? It _merged_ the state, but any objects get overwritten, because it doesn't _deeply_ merge the
+state with the object you pass into `this.setState()`. We can solve this using `Object.assign()` by merging the
+`addressInfo` object with the new data ourselves:
+
+```js
+this.setState({
+  addressInfo: Object.assign({}, this.state.addressInfo, {
+    city: 'New York City',
+  }),
+});
+```
+
+**Or**, we could do this using the proposed object spread operator in the next version of JS:
+
+```js
+this.setState({
+  addressInfo: {
+    ...this.state.addressInfo,
+    city: 'New York City',
+  }),
+});
+```
+
+Note that the object spread operator is not enabled by default in the Babel compilation of this lesson code, so we'll
+have to stick with `Object.assign()`. Just be aware that there is another way to do this!
+
+Both of these would result in the state updating to this shape:
+
+```
+{
+  theme: 'blue',
+  addressInfo: {
+    street: null,
+    number: null,
+    city: 'New York City',
+    country: null
+  },
+}
+```
+
+Perfect! Just what we needed.
+
+## Setting state is not synchronous
+One thing to keep in mind is that setting state is _not_ synchronous. For all intents and purposes, it might seem that
+way, since our components update right away. State updates, however, are _batched_ internally and then executed
+simultaneously whenever React feels it's appropriate. This might result in some unexpected behavior. Going back to our
+`ClickityClick` component above, let's log the state after we've set it using `this.setState()`:
+
+```js
+handleClick() {
+  this.setState({
+    hasBeenClicked: true,
+  });
+  console.log(this.state.hasBeenClicked); // prints false
+}
+```
+
+The console output says `false`... but we just set it to `true`! What is this madness?
+
+State changes, however instant they might appear, happen _asynchronously_. If we want to access our new state after it
+has been updated, we can optionally add a callback as a second argument to the `this.setState()` function. This callback
+will fire once the state has been updated, ensuring that `this.state` is now the new, shiny updated state. In code:
+
+```js
+handleClick() {
+  this.setState(
+    { hasBeenClicked: true },
+    function () {
+      console.log(this.state.hasBeenClicked); // prints true
+    }
+  );
+}
+```
+
+## State changes vs. prop changes
+![Not the same thing!](http://4.bp.blogspot.com/-YpCHzw3WdTo/UzNBI3BzYKI/AAAAAAAAJoY/S34pUkXKhUU/s1600/aaa.png)
+
+Just a quick sidenote, but it's important to distinguish changes in state and changes in props. Changes in state and/or
```

Comment:
I might rephrase this a bit. This seems to be an actually quite important (even crucial), point to make. "Just a quick sidenote" kind of downplays its importance. Or if you want to keep this phrasing, I'd do instead: "Just a quick sidenote, but..." -> "Just a quick sidenote: It's important to distinguish changes..." (since "but" is not really the logical relation that you are expressing here...).